core/utils/log.hpp

```diff
@@ -66,56 +68,88 @@ enum level_t {
   IRL_TRACE
 };
 
+//////////////////////////////////////////////////////////////////////////////
+/// @brief log appender callback
+//////////////////////////////////////////////////////////////////////////////
+typedef void  (*log_appender_callback_t)(void* context, const char* function, const char* file, int line,
+                              level_t level, const char* message, 
+                              size_t message_len);
+
 IRESEARCH_API bool enabled(level_t level);
-IRESEARCH_API FILE* output(level_t level);
+
+// Backward compatible fd-appender control functions
 IRESEARCH_API void output(level_t level, FILE* out); // nullptr == /dev/null
 IRESEARCH_API void output_le(level_t level, FILE* out); // nullptr == /dev/null
+// Custom appender control functions
+IRESEARCH_API void output(level_t level, log_appender_callback_t appender, void* context); // nullptr == appender -> log level disabled
+IRESEARCH_API void output_le(level_t level, log_appender_callback_t appender, void* context); // nullptr == appender -> log level disabled
+
+IRESEARCH_API void log(const char* function, const char* file, int line,
+                       level_t level, const char* message, size_t len);
+
 IRESEARCH_API void stack_trace(level_t level);
 IRESEARCH_API void stack_trace(level_t level, const std::exception_ptr& eptr);
 IRESEARCH_API irs::logger::level_t stack_trace_level(); // stack trace output level
 IRESEARCH_API void stack_trace_level(level_t level); // stack trace output level
-IRESEARCH_API std::ostream& stream(level_t level);
+
+struct log_vararg_wrapper {
```

Comment:
why `struct`, but not a single func?
move to `cpp`?

Reply:
That`s for shared builds. Avoids exporting unnecessary wrapper function